gen/build/src/lib.rs

```diff
@@ -179,6 +179,8 @@ impl Project {
             TargetDir::Unknown => scratch::path("cxxbridge"),
         };
 
+        println!("cargo:rerun-if-env-changed=GENERATED_HEADER_DIR");
```

Comment:
This use of an environment variable is not compatible with Cargo's expectations around build script behavior. As written, this definitely isn't something we can land because it breaks any build script that does not already contain its own accurate "cargo:rerun-if-…" lines. If cxx-build prints this line, cargo will not rerun the build script for *any* other relevant reason, for example even source file changes in the file holding the #\[cxx::bridge\]. Meanwhile not printing this line is also not correct, because build scripts would end up not seeing changes to this environment variable.

Would you be able to propose some other ways to handle what you need in cxx-qt?

Reply:
>  breaks any build script that does not already contain its own accurate "cargo:rerun-if-…" lines

Wouldn't such a build script kinda be broken already? The example in cxx-build's documentation clearly shows the build script should have `cargo:rerun-if-changed` lines for the bridge files and [Cargo's documentation](https://doc.rust-lang.org/cargo/reference/build-scripts.html#change-detection) recommends this:

> it is recommended that every build script emit at least one of the rerun-if instructions

On the other hand, since it is already the build script's responsibility to print `rerun-if-changed` for the bridge files, it would be consistent to leave `rerun-if-env-changed=GENERATED_HEADER_DIR` to the build script...

Reply:
It would generally not already be kinda broken. `rerun-if-changed` is optional and Cargo has the sane default behavior of rerunning if any file in the project has changed. For example just from the first page of https://crates.io/crates/cxx-build/reverse_dependencies, here are several build scripts that are not broken today but would be broken by adding a `rerun-if-env-changed` into cxx-build:

- https://github.com/MaterializeInc/rust-protobuf-native/blob/protobuf-native-v0.2.1+3.19.1/protobuf-native/build.rs
- https://github.com/tsurucapital/quadprogpp-rs/blob/v0.1.0/quadprogpp-sys/build.rs
- https://github.com/risc0/risc0/blob/v0.11.1/risc0/core/build.rs (and 4 other build.rs in that repo)

Reply:
Thinking about this further, either this must be the build script's responsibility, or cxx-build would need to only print this for bridges with `extern Rust` sections. Crates that only have `extern C++` sections would not need this.

Reply:
Even among crates with `extern Rust` sections, they might not be designed to be used with build systems besides Cargo. For example, the `extern Rust` could only be used to interact with C++ files within the crate that are compiled with cc by the Cargo build script. In that case, `rerun-if-changed-env=GENERATED_HEADER_DIR` would be wrong. So the decision whether to print `rerun-if-changed-env` must be left to the downstream build script.

The primary use case for this is for Rust libraries written to be linked into a specific C++ application. I doubt people will want to use this to create C++ bindings for general purpose Rust libraries, though that would technically be possible.